// @blnkfinance/blnk-typescript/dist/tests/unit/blnk/endpoints/identity.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable n/no-unpublished-import */
const tap_1 = __importDefault(require("tap"));
const identity_1 = require("../../../../src/blnk/endpoints/identity");
const blnkClientMocks_1 = require("../../../mocks/blnkClientMocks");
const httpClient_1 = require("../../../../src/blnk/utils/httpClient");
tap_1.default.test(`Identity`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    t.test(`Create Organization Identity`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const identity = new identity_1.Identity(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            category: `test`,
            identity_type: `organization`,
            city: `test`,
            country: `test`,
            email_address: `test@test.com`,
            organization_name: `test org`,
            state: `test`,
            post_code: `test`,
            street: `test`,
            phone_number: `1234567890`,
        };
        const response = yield identity.create(data);
        childTest.match(capturedRequest.args(), [[`identities`, data, `POST`]]);
        childTest.equal(response.status, 201);
        childTest.equal((_a = response.data) === null || _a === void 0 ? void 0 : _a.identity_type, data.identity_type);
    }));
    t.test(`It should handle missing fields if organization is selected`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const identity = new identity_1.Identity(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            category: `test`,
            identity_type: `organization`,
            city: `test`,
            country: `test`,
            email_address: `test@test.com`,
            state: `test`,
            post_code: `test`,
            street: `test`,
            phone_number: `1234567890`,
        };
        const response = yield identity.create(data);
        childTest.match(capturedRequest.args(), []);
        childTest.equal(response.data, null);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.test(`It should handle missing fields if individual is selected`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const identity = new identity_1.Identity(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            category: `test`,
            identity_type: `individual`,
            city: `test`,
            country: `test`,
            email_address: `test@test.com`,
            state: `test`,
            post_code: `test`,
            street: `test`,
            phone_number: `1234567890`,
        };
        const response = yield identity.create(data);
        childTest.match(capturedRequest.args(), []);
        childTest.equal(response.data, null);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.test(`it should handle errors thrown during creation`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(false, `Error creating identity`, 500);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const identity = new identity_1.Identity(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            category: `test`,
            identity_type: `organization`,
            city: `test`,
            country: `test`,
            email_address: `test@test.com`,
            organization_name: `test org`,
            state: `test`,
            post_code: `test`,
            street: `test`,
            phone_number: `1234567890`,
        };
        const response = yield identity.create(data);
        childTest.match(capturedRequest.args(), [[`identities`, data, `POST`]]);
        childTest.equal(response.status, 500);
        childTest.equal(response.data, null);
        childTest.end();
    }));
}));
