// @blnkfinance/blnk-typescript/dist/tests/unit/blnk/endpoints/reconciliation.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable n/no-unpublished-import */
const tap_1 = __importDefault(require("tap"));
const reconciliation_1 = require("../../../../src/blnk/endpoints/reconciliation");
const blnkClientMocks_1 = require("../../../mocks/blnkClientMocks");
const httpClient_1 = require("../../../../src/blnk/utils/httpClient");
const fs_1 = require("fs");
const form_data_1 = __importDefault(require("form-data"));
const path_1 = __importDefault(require("path"));
const filePath = path_1.default.join(__dirname, `..`, `..`, `..`, `..`, `file.csv`);
tap_1.default.test(`Reconciliation`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    t.test(`Create Matching Rule`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            criteria: [
                {
                    field: `amount`,
                    operator: `equals`,
                    allowable_drift: 0.01,
                },
            ],
            description: `Test Matching Rule`,
            name: `Test Matching Rule`,
        };
        const response = yield reconciliation.createMatchingRule(data);
        childTest.match(capturedRequest.args(), [
            [`reconciliation/matching-rules`, data, `POST`],
        ]);
        childTest.match(response.status, 201);
        childTest.end();
    }));
    t.test(`should upload file successfully when given a valid file path`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const response = yield reconciliation.upload(filePath, `Stripe`);
        childTest.type(capturedRequest.args()[0][1], form_data_1.default); //makes sure that the second parameter called in the function is FormData
        childTest.match(response.status, 201);
        childTest.end();
    }));
    t.test(`should handle error gracefully when given an invalid file path`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const filePath = `./no_file.csv`;
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(false, `File does not exist at path: ${filePath}`, 404);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const response = yield reconciliation.upload(filePath, `Stripe`);
        childTest.match(response.status, 404);
        childTest.end();
    }));
    t.test(`should upload file successfully when given a valid Readstream`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const readStream = (0, fs_1.createReadStream)(filePath);
        const response = yield reconciliation.upload(readStream, `Stripe`);
        childTest.type(capturedRequest.args()[0][1], form_data_1.default); //makes sure that the second parameter called in the function is FormData
        childTest.match(response.status, 201);
        childTest.end();
    }));
}));
tap_1.default.test(`Run Reconciliation`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    t.test(`Start Reconciliation`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            upload_id: `0987654321`,
            matching_rule_ids: [`1233455555`],
            dry_run: false,
            strategy: `one_to_many`,
            grouping_criteria: `amount`,
        };
        const response = yield reconciliation.run(data);
        childTest.match(capturedRequest.args(), [
            [`reconciliation/start`, data, `POST`],
        ]);
        childTest.match(response.status, 201);
        childTest.end();
    }));
    t.test(`should handle thrown errors gracefully`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(false, `An error occurred`, 500);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const reconciliation = new reconciliation_1.Reconciliation(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            upload_id: `0987654321`,
            matching_rule_ids: [`1233455555`],
            dry_run: false,
            strategy: `one_to_many`,
            grouping_criteria: `amount`,
        };
        const response = yield reconciliation.run(data);
        childTest.match(response.status, 500);
        childTest.end();
    }));
}));
