// @blnkfinance/blnk-typescript/dist/tests/unit/blnk/endpoints/transactions.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable n/no-unpublished-import */
const tap_1 = __importDefault(require("tap"));
const transactions_1 = require("../../../../src/blnk/endpoints/transactions");
const blnkClientMocks_1 = require("../../../mocks/blnkClientMocks");
const httpClient_1 = require("../../../../src/blnk/utils/httpClient");
tap_1.default.test(`Creates a transaction`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
    let thirdPartyRequest;
    t.beforeEach(() => {
        thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 201);
    });
    t.test(`Creates a transaction with valid data`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b, _c;
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            amount: 10000,
            currency: `USD`,
            description: `Test transaction`,
            meta_data: { company_name: `Test Company` },
            precision: 100,
            reference: `1234567890`,
        };
        const transaction = yield transactions.create(data);
        childTest.match(capturedRequest.args(), [[`transactions`, data, `POST`]]);
        childTest.equal((_a = transaction.data) === null || _a === void 0 ? void 0 : _a.amount, data.amount);
        childTest.equal((_b = transaction.data) === null || _b === void 0 ? void 0 : _b.currency, data.currency);
        childTest.equal((_c = transaction.data) === null || _c === void 0 ? void 0 : _c.description, data.description);
        childTest.end();
    }));
    t.test(`It should handle missing required fields`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        //we cast to any here so we can simulate a js user forgetting to put in a compulsory field, since by default Typescript should catch this during development
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const data = {
            currency: `USD`,
            description: `Test transaction`,
            meta_data: { company_name: `Test Company` },
            precision: 100,
            amount: 10000,
        };
        const response = yield transactions.create(data);
        childTest.match(capturedRequest.args(), []);
        childTest.equal(response.data, null);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.test(`it should handle thrown errors during creation`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(false, `Something went wrong`);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            amount: 10000,
            currency: `USD`,
            description: `Test transaction`,
            meta_data: { company_name: `Test Company` },
            precision: 100,
            reference: `1234567890`,
        };
        const response = yield transactions.create(data);
        childTest.match(capturedRequest.args(), [[`transactions`, data, `POST`]]);
        childTest.equal(response.data, null);
        childTest.equal(response.status, 500);
        childTest.equal(response.message, `Something went wrong`);
        childTest.end();
    }));
    t.test(`it should handle meta_data if it is not an object`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(false, `Something went wrong`);
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        //we cast to any here so we can simulate a js user forgetting to put in a compulsory field, since by default Typescript should catch this during development
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const data = {
            amount: 10000,
            currency: `USD`,
            description: `Test transaction`,
            meta_data: `Test Company`,
            precision: 100,
            reference: `1234567890`,
        };
        const response = yield transactions.create(data);
        childTest.match(capturedRequest.args(), []);
        childTest.equal(response.data, null);
        childTest.equal(response.status, 400);
        childTest.equal(response.message, `meta_data must be a valid object if provided`);
    }));
}));
tap_1.default.test(`Updates a transaction`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const mockLogger = (0, blnkClientMocks_1.createMockLogger)();
    let thirdPartyRequest;
    t.beforeEach(() => {
        thirdPartyRequest = (0, blnkClientMocks_1.createMockBlnkRequest)(true, undefined, 200);
    });
    const id = `1234`;
    t.test(`Updates a transaction status with valid data`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        const data = {
            status: `commit`,
        };
        const transaction = yield transactions.updateStatus(id, data);
        childTest.match(capturedRequest.args(), [
            [`transactions/inflight/${id}`, data, `PUT`],
        ]);
        childTest.equal(transaction.status, 200);
        childTest.end();
    }));
    t.test(`Updates fails for a transaction status with invalid data`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const capturedRequest = childTest.captureFn(thirdPartyRequest);
        const transactions = new transactions_1.Transactions(capturedRequest, mockLogger, httpClient_1.FormatResponse);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const data = {
            status: `commit`,
            meta_data: `Test Company`,
        };
        const transaction = yield transactions.updateStatus(id, data);
        childTest.match(capturedRequest.args(), []);
        childTest.equal(transaction.data, null);
        childTest.equal(transaction.status, 400);
    }));
}));
