// nestjs-ddd-cli/dist/utils/dependency.utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPackageInstalledGlobally = isPackageInstalledGlobally;
exports.getLatestVersion = getLatestVersion;
exports.isLatestVersion = isLatestVersion;
exports.updatePackageGlobally = updatePackageGlobally;
exports.isNestJSCliInstalled = isNestJSCliInstalled;
exports.installNestJSCli = installNestJSCli;
exports.createNestJSProject = createNestJSProject;
exports.installDependencies = installDependencies;
exports.checkOutdatedDependencies = checkOutdatedDependencies;
exports.updateDependencies = updateDependencies;
exports.getCurrentPackageVersion = getCurrentPackageVersion;
exports.checkForCliUpdate = checkForCliUpdate;
const child_process_1 = require("child_process");
const util_1 = require("util");
const fs = __importStar(require("fs-extra"));
const path = __importStar(require("path"));
const chalk_1 = __importDefault(require("chalk"));
const execAsync = (0, util_1.promisify)(child_process_1.exec);
/**
 * Check if a package is installed globally
 */
async function isPackageInstalledGlobally(packageName) {
    try {
        await execAsync(`npm list -g ${packageName} --depth=0`);
        return true;
    }
    catch (error) {
        return false;
    }
}
/**
 * Get the latest version of a package from npm
 */
async function getLatestVersion(packageName) {
    try {
        const { stdout } = await execAsync(`npm view ${packageName} version`);
        return stdout.trim();
    }
    catch (error) {
        throw new Error(`Failed to get latest version of ${packageName}: ${error.message}`);
    }
}
/**
 * Check if the current version is the latest version
 */
async function isLatestVersion(packageName, currentVersion) {
    try {
        const latestVersion = await getLatestVersion(packageName);
        return currentVersion === latestVersion;
    }
    catch (error) {
        console.warn(chalk_1.default.yellow(`Warning: Could not check for updates: ${error.message}`));
        return true; // Assume it's the latest version if we can't check
    }
}
/**
 * Update a package globally
 */
async function updatePackageGlobally(packageName) {
    try {
        console.log(chalk_1.default.blue(`Updating ${packageName} globally...`));
        await execAsync(`npm install -g ${packageName}@latest`);
        console.log(chalk_1.default.green(`✅ ${packageName} updated successfully!`));
    }
    catch (error) {
        throw new Error(`Failed to update ${packageName}: ${error.message}`);
    }
}
/**
 * Check if NestJS CLI is installed
 */
async function isNestJSCliInstalled() {
    return isPackageInstalledGlobally('@nestjs/cli');
}
/**
 * Install NestJS CLI globally
 */
async function installNestJSCli() {
    try {
        console.log(chalk_1.default.blue('Installing @nestjs/cli globally...'));
        await execAsync('npm install -g @nestjs/cli');
        console.log(chalk_1.default.green('✅ @nestjs/cli installed successfully!'));
    }
    catch (error) {
        throw new Error(`Failed to install @nestjs/cli: ${error.message}`);
    }
}
/**
 * Create a new NestJS project
 */
async function createNestJSProject(projectName, options = {}) {
    try {
        // Check if @nestjs/cli is installed
        const isInstalled = await isNestJSCliInstalled();
        if (!isInstalled) {
            await installNestJSCli();
        }
        console.log(chalk_1.default.blue(`Creating new NestJS project: ${projectName}...`));
        // Build the command with options
        let command = `nest new ${projectName} --skip-git --package-manager npm`;
        if (options.directory) {
            command += ` --directory ${options.directory}`;
        }
        if (options.skipInstall) {
            command += ' --skip-install';
        }
        await execAsync(command);
        console.log(chalk_1.default.green(`✅ NestJS project ${projectName} created successfully!`));
    }
    catch (error) {
        throw new Error(`Failed to create NestJS project: ${error.message}`);
    }
}
/**
 * Install dependencies in a project
 */
async function installDependencies(projectPath, dependencies, dev = false) {
    try {
        const flag = dev ? '--save-dev' : '--save';
        console.log(chalk_1.default.blue(`Installing ${dev ? 'dev ' : ''}dependencies: ${dependencies.join(', ')}...`));
        await execAsync(`npm install ${flag} ${dependencies.join(' ')}`, { cwd: projectPath });
        console.log(chalk_1.default.green('✅ Dependencies installed successfully!'));
    }
    catch (error) {
        throw new Error(`Failed to install dependencies: ${error.message}`);
    }
}
/**
 * Check for outdated dependencies in a project
 */
async function checkOutdatedDependencies(projectPath) {
    try {
        const { stdout } = await execAsync('npm outdated --json', { cwd: projectPath });
        return JSON.parse(stdout || '{}');
    }
    catch (error) {
        // npm outdated returns exit code 1 if there are outdated packages
        if (error.stdout) {
            try {
                return JSON.parse(error.stdout);
            }
            catch {
                return {};
            }
        }
        return {};
    }
}
/**
 * Update dependencies in a project
 */
async function updateDependencies(projectPath, dependencies) {
    try {
        console.log(chalk_1.default.blue(`Updating dependencies: ${dependencies.join(', ')}...`));
        await execAsync(`npm update ${dependencies.join(' ')}`, { cwd: projectPath });
        console.log(chalk_1.default.green('✅ Dependencies updated successfully!'));
    }
    catch (error) {
        throw new Error(`Failed to update dependencies: ${error.message}`);
    }
}
/**
 * Get the current package version from package.json
 */
function getCurrentPackageVersion() {
    try {
        const packageJsonPath = path.resolve(__dirname, '../../package.json');
        const packageJson = fs.readJsonSync(packageJsonPath);
        return packageJson.version;
    }
    catch (error) {
        throw new Error(`Failed to get current package version: ${error.message}`);
    }
}
/**
 * Check if the CLI needs to be updated
 */
async function checkForCliUpdate() {
    const currentVersion = getCurrentPackageVersion();
    try {
        const latestVersion = await getLatestVersion('nestjs-ddd-cli');
        const needsUpdate = currentVersion !== latestVersion;
        return { needsUpdate, latestVersion, currentVersion };
    }
    catch (error) {
        console.warn(chalk_1.default.yellow(`Warning: Could not check for CLI updates: ${error.message}`));
        return { needsUpdate: false, latestVersion: currentVersion, currentVersion };
    }
}
