// @blnkfinance/blnk-typescript/dist/tests/mocks/blnkClientMocks.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockBlnkRequest = exports.ledgerId = exports.createMockServices = exports.createMockBlnkClientOptions = exports.createMockLogger = void 0;
const mockTransactions_1 = require("./mockTransactions");
/**
 * Manually create a mock logger that adheres to the BlnkLogger interface
 */
const createMockLogger = () => {
    return {
        info: (message, ...meta) => {
            console.log(`Mock Info: ${message}`, ...meta);
        },
        error: (message, ...meta) => {
            console.log(`Mock Error: ${message}`, ...meta);
        },
        debug: (message, ...meta) => {
            console.log(`Mock Debug: ${message}`, ...meta);
        },
    };
};
exports.createMockLogger = createMockLogger;
/**
 * Manually create mock BlnkClientOptions, including the mock logger
 */
const createMockBlnkClientOptions = () => {
    return {
        baseUrl: `http://mock-api.com`,
        timeout: 5000,
        logger: (0, exports.createMockLogger)(),
    };
};
exports.createMockBlnkClientOptions = createMockBlnkClientOptions;
//export const createMockRequest = <T, R>(): BlnkRequest => t.createMock();
const createMockServices = () => ({
    Ledgers: mockTransactions_1.TransactionMock.Transactions,
    LedgerBalances: mockTransactions_1.TransactionMock.Transactions,
    Transactions: mockTransactions_1.TransactionMock.Transactions,
});
exports.createMockServices = createMockServices;
// Mock ApiResponse implementation
const mockApiResponse = (data, status = 200, message = `Success`) => {
    return {
        status,
        message,
        data,
    };
};
exports.ledgerId = `123456`;
// Mock BlnkRequest implementation
// Factory function for creating mockBlnkRequest
const createMockBlnkRequest = (success, throwError = undefined, status = 200) => {
    return (endpoint, data, method, headerOptions) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            if (throwError !== undefined) {
                throw new Error(throwError);
            }
            // Simulate success or error based on the success parameter
            if (success) {
                const mockData = Object.assign(Object.assign({}, data), { ledger_id: exports.ledgerId });
                return mockApiResponse(mockData, status);
            }
            else {
                return mockApiResponse(null, 500, `Internal Server Error`);
            }
        }
        catch (error) {
            throw new Error(throwError);
        }
    });
};
exports.createMockBlnkRequest = createMockBlnkRequest;
