// @blnkfinance/blnk-typescript/dist/tests/mocks/fetchMock.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchFailMock = exports.fetchMock = void 0;
/* eslint-disable n/no-unpublished-import */
// src/mocks/fetchMock.ts
const tap_1 = __importDefault(require("tap"));
const node_fetch_1 = __importStar(require("node-fetch"));
exports.fetchMock = tap_1.default.createMock({ fetch: node_fetch_1.default }, {
    fetch: (
    // eslint-disable-next-line n/no-unsupported-features/node-builtins
    input, init) => __awaiter(void 0, void 0, void 0, function* () {
        return {
            ok: true,
            status: 200,
            json: () => __awaiter(void 0, void 0, void 0, function* () {
                return ({
                    message: `Success`,
                });
            }),
            text: () => __awaiter(void 0, void 0, void 0, function* () { return `Mock text response`; }),
            headers: new node_fetch_1.Headers(init === null || init === void 0 ? void 0 : init.headers), // Simulate headers if needed
            statusText: `OK`,
            url: input.toString(),
            redirected: false,
            type: `basic`,
            body: null,
            bodyUsed: false,
        }; // Cast to Response
    }),
});
exports.fetchFailMock = tap_1.default.createMock({ fetch: node_fetch_1.default }, {
    fetch: (
    // eslint-disable-next-line n/no-unsupported-features/node-builtins
    input, init) => __awaiter(void 0, void 0, void 0, function* () {
        return {
            ok: false,
            status: 500,
            json: () => __awaiter(void 0, void 0, void 0, function* () {
                return ({
                    message: `Failed`,
                });
            }),
            text: () => __awaiter(void 0, void 0, void 0, function* () { return `Failed`; }),
            headers: new node_fetch_1.Headers(init === null || init === void 0 ? void 0 : init.headers), // Simulate headers if needed
            statusText: `Failed`,
            url: input.toString(),
            redirected: false,
            type: `basic`,
            body: null,
            bodyUsed: false,
        }; // Cast to Response
    }),
});
