// @blnkfinance/blnk-typescript/dist/tests/mocks/mockTransactions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionMock = void 0;
exports.createDummyTransactionResponse = createDummyTransactionResponse;
/* eslint-disable n/no-unpublished-import */
const transactions_1 = require("../../src/blnk/endpoints/transactions");
const tap_1 = __importDefault(require("tap"));
class MockTransaction {
    constructor(request, logger, formatResponse) {
        this.request = request;
        this.logger = logger;
        this.formatResponse = formatResponse;
    }
    create(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                status: 200,
                data: Object.assign(Object.assign({}, data), createDummyTransactionResponse()),
                message: `Success`,
            };
        });
    }
    updateStatus(id, update) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                data: Object.assign(Object.assign(Object.assign(Object.assign({}, createDummyTransactionResponse()), { transaction_id: id }), update), { status: `COMMIT` }),
                message: `Success`,
                status: 200,
            };
        });
    }
    refund(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                data: Object.assign(Object.assign({}, createDummyTransactionResponse()), { transaction_id: id }),
                message: `Success`,
                status: 200,
            };
        });
    }
}
// Mock class for Transactions
exports.TransactionMock = tap_1.default.createMock({ Transactions: transactions_1.Transactions }, {
    Transactions: MockTransaction,
});
function createDummyTransactionResponse() {
    return {
        transaction_id: `txn_` + Math.random().toString(36).substr(2, 9),
        amount: 1000,
        precision: 2,
        precise_amount: 100000,
        rate: 0,
        reference: `REF12345`,
        description: `Sample transaction description`,
        currency: `USD`,
        status: `INFLIGHT`,
        source: `source_12345`,
        destination: `destination_67890`,
        sources: [
            {
                identifier: `account1`,
                distribution: `left`,
            },
        ],
        created_at: new Date(),
        meta_data: {},
    };
}
