// nestjs-ddd-cli/dist/commands/init-project.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initProject = initProject;
const path = __importStar(require("path"));
const inquirer_1 = __importDefault(require("inquirer"));
const chalk_1 = __importDefault(require("chalk"));
const dependency_utils_1 = require("../utils/dependency.utils");
async function initProject(projectName, options) {
    try {
        // First, check for CLI updates if not skipped
        if (!options.skipUpdate) {
            const { needsUpdate, latestVersion, currentVersion } = await (0, dependency_utils_1.checkForCliUpdate)();
            if (needsUpdate) {
                console.log(chalk_1.default.yellow(`You are using nestjs-ddd-cli version ${currentVersion}, but version ${latestVersion} is available.`));
                const { shouldUpdate } = await inquirer_1.default.prompt([
                    {
                        type: 'confirm',
                        name: 'shouldUpdate',
                        message: 'Would you like to update the CLI before continuing?',
                        default: true,
                    },
                ]);
                if (shouldUpdate) {
                    await (0, dependency_utils_1.updatePackageGlobally)('nestjs-ddd-cli');
                    console.log(chalk_1.default.green('CLI updated successfully! Please run your command again.'));
                    process.exit(0);
                }
            }
        }
        // Determine the project directory
        const projectDir = options.path ? path.resolve(options.path, projectName) : path.resolve(process.cwd(), projectName);
        // Create a new NestJS project
        await (0, dependency_utils_1.createNestJSProject)(projectName, {
            directory: options.path,
            skipInstall: options.skipInstall,
        });
        // If withDdd option is enabled, install DDD-related dependencies
        if (options.withDdd) {
            const dependencies = ['@nestjs/cqrs', 'class-validator', 'class-transformer'];
            await (0, dependency_utils_1.installDependencies)(projectDir, dependencies);
            // Create DDD folder structure
            console.log(chalk_1.default.blue('Setting up DDD folder structure...'));
            // We'll use the generate-module command to create the initial module structure
            // This will be implemented in the main CLI entry point
            console.log(chalk_1.default.green('✅ DDD folder structure set up successfully!'));
        }
        console.log(chalk_1.default.green(`\n✅ Project ${projectName} initialized successfully!`));
        console.log(chalk_1.default.blue(`\nNext steps:`));
        console.log(chalk_1.default.blue(`  1. cd ${projectName}`));
        console.log(chalk_1.default.blue(`  2. npm run start:dev`));
        if (options.withDdd) {
            console.log(chalk_1.default.blue(`\nTo generate DDD components, use:`));
            console.log(chalk_1.default.blue(`  ddd generate module <module-name>`));
            console.log(chalk_1.default.blue(`  ddd generate entity <entity-name> --module <module-name>`));
            console.log(chalk_1.default.blue(`  ddd generate usecase <usecase-name> --module <module-name>`));
        }
    }
    catch (error) {
        console.error(chalk_1.default.red('Error:'), error.message);
        process.exit(1);
    }
}
