// nestjs-ddd-cli/dist/commands/update-deps.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateDeps = updateDeps;
const inquirer_1 = __importDefault(require("inquirer"));
const chalk_1 = __importDefault(require("chalk"));
const dependency_utils_1 = require("../utils/dependency.utils");
async function updateDeps(options = {}) {
    try {
        const projectPath = options.path || process.cwd();
        console.log(chalk_1.default.blue('Checking for outdated dependencies...'));
        const outdatedDeps = await (0, dependency_utils_1.checkOutdatedDependencies)(projectPath);
        const outdatedPackages = Object.keys(outdatedDeps);
        if (outdatedPackages.length === 0) {
            console.log(chalk_1.default.green('✅ All dependencies are up to date!'));
            return;
        }
        console.log(chalk_1.default.yellow(`Found ${outdatedPackages.length} outdated dependencies:`));
        for (const pkg of outdatedPackages) {
            const info = outdatedDeps[pkg];
            if (info) {
                const { current, latest, type } = info;
                console.log(chalk_1.default.yellow(`  ${pkg}: ${current} → ${latest} (${type})`));
            }
        }
        let packagesToUpdate = [];
        if (options.all) {
            packagesToUpdate = outdatedPackages;
        }
        else {
            const { selectedPackages } = await inquirer_1.default.prompt([
                {
                    type: 'checkbox',
                    name: 'selectedPackages',
                    message: 'Select packages to update:',
                    choices: outdatedPackages.map(pkg => {
                        const info = outdatedDeps[pkg];
                        return {
                            name: info ? `${pkg}: ${info.current} → ${info.latest}` : pkg,
                            value: pkg,
                            checked: true,
                        };
                    }),
                },
            ]);
            packagesToUpdate = selectedPackages;
        }
        if (packagesToUpdate.length === 0) {
            console.log(chalk_1.default.yellow('No packages selected for update.'));
            return;
        }
        await (0, dependency_utils_1.updateDependencies)(projectPath, packagesToUpdate);
    }
    catch (error) {
        console.error(chalk_1.default.red('Error:'), error.message);
        process.exit(1);
    }
}
