// @blnkfinance/blnk-typescript/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlnkInit = BlnkInit;
const balanceMonitors_1 = require("./blnk/endpoints/balanceMonitors");
const baseBlnkClient_1 = require("./blnk/endpoints/baseBlnkClient");
const identity_1 = require("./blnk/endpoints/identity");
const ledgerBalances_1 = require("./blnk/endpoints/ledgerBalances");
const ledgers_1 = require("./blnk/endpoints/ledgers");
const reconciliation_1 = require("./blnk/endpoints/reconciliation");
const search_1 = require("./blnk/endpoints/search");
const transactions_1 = require("./blnk/endpoints/transactions");
const httpClient_1 = require("./blnk/utils/httpClient");
const logger_1 = require("./blnk/utils/logger");
const node_fetch_1 = __importDefault(require("node-fetch"));
//import {BlnkInitFn} from "./types/general";
// Export a function to initialize the SDK with default logger handling
function BlnkInit(apiKey, options) {
    if (options.logger === undefined) {
        options.logger = logger_1.CustomLogger;
    }
    return new baseBlnkClient_1.Blnk(apiKey, options, {
        Ledgers: ledgers_1.Ledgers,
        LedgerBalances: ledgerBalances_1.LedgerBalances,
        Transactions: transactions_1.Transactions,
        BalanceMonitor: balanceMonitors_1.BalanceMonitor,
        Reconciliation: reconciliation_1.Reconciliation,
        Search: search_1.Search,
        Identity: identity_1.Identity,
    }, httpClient_1.FormatResponse, node_fetch_1.default);
}
//module.exports = BlnkInit as BlnkInitFn & {default: BlnkInitFn};
//export default BlnkInit;
