// @blnkfinance/blnk-typescript/dist/src/blnk/endpoints/baseBlnkClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blnk = void 0;
const logger_1 = require("../utils/logger");
const form_data_1 = __importDefault(require("form-data"));
/**
 * Blnk class for interacting with the Blnk API services.
 *
 * Makes requests to specified endpoints using provided data and methods.
 * Caches initialized services for efficient usage.
 *
 * @example
 * const services = {
      Ledgers,
      LedgerBalances,
      Transactions,
      BalanceMonitor,
      Reconciliation,
      Search,
      Identity,
    }
 * const blnk = new Blnk(apiKey, options, services, formatResponse, thirdPartyRequest);
 * const ledgersService = blnk.Ledgers;
 * const ledgersData = await ledgersService.getLedgersData();
 */
class Blnk {
    constructor(apiKey, options, services, formatResponse, thirdPartyRequest) {
        this.serviceInstances = {}; // Cache initialized services
        if (!options.baseUrl) {
            throw new Error(`baseUrl is required for self-hosted Blnk SDK.`);
        }
        //make sure baseUrl ends with "/"
        if (!options.baseUrl.endsWith(`/`)) {
            options.baseUrl = options.baseUrl + `/`;
        }
        this.apiKey = apiKey;
        const { logger } = options, restOptions = __rest(options, ["logger"]);
        this.options = Object.assign({ timeout: 3000 }, restOptions);
        if (logger === undefined) {
            this.logger = console;
        }
        else {
            this.logger = logger;
        }
        this.services = services;
        this.formatResponse = formatResponse;
        this.thirdPartyRequest = thirdPartyRequest;
    }
    /**
     * Makes a request to the specified endpoint using the provided data and method.
     *
     * @param endpoint The endpoint to send the request to. should be "endpoint" to '/' at the beginning.
     * @param data The data to be sent with the request.
     * @param method The HTTP method for the request (POST, GET, PUT, DELETE).
     * @returns A Promise that resolves to an ApiResponse object containing the response data or null in case of an error.
     */
    request(endpoint, data, method, headerOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = Object.assign({ "content-type": `application/json`, "X-Blnk-Key": this.apiKey }, headerOptions);
            try {
                const response = yield this.thirdPartyRequest(`${this.options.baseUrl}${endpoint}`, {
                    method,
                    headers,
                    body: data
                        ? data instanceof form_data_1.default
                            ? data
                            : JSON.stringify(data)
                        : undefined,
                    timeout: this.options.timeout,
                });
                if (!response.ok) {
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    const errorResult = yield response.json();
                    this.logger.error(`Request to ${endpoint} failed with status ${response.status}.`);
                    return this.formatResponse(response.status, response.statusText, errorResult);
                }
                const jsonResponse = (yield response.json());
                return this.formatResponse(response.status, `Success`, jsonResponse);
            }
            catch (error) {
                this.logger.error(`Request failed`, { endpoint, error });
                return (0, logger_1.HandleError)(error, this.logger, this.formatResponse, `${this.request.name}`);
            }
        });
    }
    /**
     * Retrieves a registered service by name and returns an instance of that service.
     * Throws an error if the service is not registered.
     *
     * @param serviceName - The name of the service to retrieve.
     * @returns An instance of the requested service.
     *
     * @example
     * // Retrieve the 'Ledgers' service
     * const ledgersService = getService<Ledgers>('Ledgers');
     *
     * @example
     * // Retrieve the 'Transactions' service
     * const transactionsService = getService<Transactions>('Transactions');
     */
    getService(serviceName) {
        if (!this.services[serviceName]) {
            throw new Error(`Service ${serviceName} is not registered`);
        }
        if (!this.serviceInstances[serviceName]) {
            this.serviceInstances[serviceName] = new this.services[serviceName](this.request.bind(this), this.logger, this.formatResponse);
        }
        return this.serviceInstances[serviceName];
    }
    get Ledgers() {
        return this.getService(`Ledgers`);
    }
    get LedgerBalances() {
        return this.getService(`LedgerBalances`);
    }
    get Transactions() {
        return this.getService(`Transactions`);
    }
    get BalanceMonitor() {
        return this.getService(`BalanceMonitor`);
    }
    get Reconciliation() {
        return this.getService(`Reconciliation`);
    }
    get Search() {
        return this.getService(`Search`);
    }
    get Identity() {
        return this.getService(`Identity`);
    }
    get getApiKey() {
        return this.apiKey;
    }
}
exports.Blnk = Blnk;
