// @blnkfinance/blnk-typescript/dist/src/blnk/endpoints/reconciliation.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reconciliation = void 0;
const logger_1 = require("../utils/logger");
const fs_1 = __importDefault(require("fs"));
const reconciliationValidator_1 = require("../utils/validators/reconciliationValidator");
const form_data_1 = __importDefault(require("form-data"));
/**
 * Represents a class for handling reconciliation operations.
 * see @link https://docs.blnkfinance.com/reconciliations/overview for more details
 * @constructor
 * @param {BlnkRequest} request - Function for making API requests.
 * @param {BlnkLogger} logger - Logger for handling errors and logs.
 * @param {FormatResponseType} formatResponse - Function for formatting API response.
 * @method upload - Uploads a file for reconciliation.
 * @returns {Promise<ApiResponse<ReconciliationUploadResp | null>>} - The API response after uploading.
 * @method createMatchingRule - Creates a matching rule for reconciliation.
 * @param {Matcher} data - Data containing the matching rule details.
 * @returns {Promise<ApiResponse<RunReconResp | null>>} - The API response after creating the matching rule.
 * @method run - Initiates a reconciliation run.
 * @param {RunReconData} data - Data required to start the reconciliation run.
 * @returns {Promise<ApiResponse<RunReconResp | null>>} - The API response after starting the reconciliation run.
 * @example
 * const reconciliation = new Reconciliation(requestFunction, loggerInstance, formatResponseFunction);
 * const uploadResponse = await reconciliation.upload('file.txt', 'sourceA');
 * const ruleResponse = await reconciliation.createMatchingRule(matcherData);
 * const runResponse = await reconciliation.run(runData);
 */
class Reconciliation {
    constructor(request, logger, formatResponse) {
        this.request = request;
        this.logger = logger;
        this.formatResponse = formatResponse;
    }
    /**
     * Uploads a file for reconciliation.
     * see @link https://docs.blnkfinance.com/reconciliations/external-data for more details on the requirements for the file
     *
     * @param {string | ReadStream} fileInput - The file to upload, either a file path or a ReadStream.
     * @param {string} source - The source of the file can be any name you want but should describe the actual source e.g 'local', 'stripe', 'internal' etc.
     * @returns A promise that resolves to the reconciliation upload response(the upload_id returned should be stored as it will be needed when running reconciliation).
     *
     * @example
     * upload('path/to/file.txt', 'sourceA');
     * @example
     * upload(fs.createReadStream('path/to/file.txt'), 'sourceB');
     */
    upload(fileInput, source) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const formData = new form_data_1.default();
                // Determine if the input is a file path (string) or a ReadStream
                if (typeof fileInput === `string`) {
                    // If a string is passed, assume it's a file path and create a stream
                    //check if file path exists
                    if (!fs_1.default.existsSync(fileInput)) {
                        return this.formatResponse(404, `File does not exist at path: ${fileInput}`, null);
                    }
                    //get the buf
                    formData.append(`file`, fs_1.default.createReadStream(fileInput));
                }
                else {
                    // If a stream is passed, use it directly
                    //check if readstream is valid
                    if (!fileInput.readable) {
                        return this.formatResponse(400, `Invalid read stream provided`, null);
                    }
                    formData.append(`file`, fileInput);
                }
                formData.append(`source`, source);
                const response = yield this.request(`reconciliation/upload`, formData, `POST`, {
                    "content-type": `multipart/form-data;boundary=${formData.getBoundary()}`,
                });
                return response;
            }
            catch (error) {
                return (0, logger_1.HandleError)(error, this.logger, this.formatResponse, this.upload.name);
            }
        });
    }
    /**
     * Asynchronously creates a matching rule using the provided data.
     * see @link https://docs.blnkfinance.com/reconciliations/matching-rules for more details.
     *
     * Validates the data using the internal organization's Matcher validator.
     * If validation fails, returns a formatted response with a 400 status code.
     * Sends a POST request to the 'reconciliation/matching-rules' endpoint with the data.
     * Returns the response data type from the endpoint query.
     * Handles any errors that occur during the process using the internal organization's error handler.
     *
     * @param data - The Matcher object containing name, description, and criteria.
     * @returns The response data type from the matching rule creation endpoint(the matching rule id returned should be saved as it will be needed when running reconciliation).
     *
     * @example
     * const newMatchingRule = {
     *   name: "Rule 1",
     *   description: "Description of Rule 1",
     *   criteria: [{ field: "field1", operator: "equals", allowable_drift: 0.1 }]
     * };
     * const createdRule = await createMatchingRule(newMatchingRule);
     */
    createMatchingRule(data) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const validatorResponse = yield (0, reconciliationValidator_1.ValidateMatcher)(data);
                if (validatorResponse) {
                    return this.formatResponse(400, validatorResponse, null);
                }
                const response = yield this.request(`reconciliation/matching-rules`, data, `POST`);
                return response; //query endpoint and get response data type
            }
            catch (error) {
                return (0, logger_1.HandleError)(error, this.logger, this.formatResponse, this.createMatchingRule.name);
            }
        });
    }
    /**
     * Asynchronously runs a reconciliation process with the provided data.
     * see @link https://docs.blnkfinance.com/reconciliations/strategies for more details.
     *
     * @param data - The data needed to start the reconciliation process.
     * @returns A promise that resolves with the reconciliation response.
     * @throws If an error occurs during the reconciliation process, an error response is returned.
     *
     * @example
     * const reconciliationData: RunReconData = {
     *   upload_id: "123",
     *   strategy: "specific_strategy",
     *   dry_run: false,
     *   grouping_criteria: "criteria_field",
     *   matching_rule_ids: ["rule_id_1", "rule_id_2"]
     * };
     * const reconciliationResponse = await reconciliationService.run(reconciliationData);
     */
    run(data) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield this.request(`reconciliation/start`, data, `POST`);
                return response;
            }
            catch (error) {
                return (0, logger_1.HandleError)(error, this.logger, this.formatResponse, this.run.name);
            }
        });
    }
}
exports.Reconciliation = Reconciliation;
