// @blnkfinance/blnk-typescript/dist/src/blnk/endpoints/search.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Search = void 0;
const logger_1 = require("../utils/logger");
/**
 * Represents a Search class that handles searching functionality.
 * see @link https://docs.blnkfinance.com/search/overview for more details.
 * @constructor
 * @param {BlnkRequest} request - The request function for API calls.
 * @param {BlnkLogger} logger - The logger for handling logs.
 * @param {FormatResponseType} formatResponse - The function for formatting API responses.
 * @method search - Performs a search operation based on the provided data and service type.
 * @param {SearchParams} data - The search parameters.
 * @param {`ledgers` | `transactions` | `balances`} service - The type of service to search within.
 * @returns {Promise<SearchResponse>} - A promise that resolves to the search response.
 * @example
 * const search = new Search(requestFunction, loggerInstance, formatResponseFunction);
 * const searchData = { q: 'search query', page: 1, per_page: 10 };
 * const searchService = 'ledgers';
 * const result = await search.search(searchData, searchService);
 */
class Search {
    constructor(request, logger, formatResponse) {
        this.request = request;
        this.logger = logger;
        this.formatResponse = formatResponse;
    }
    search(data, service) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (!data.q) {
                    return this.formatResponse(400, `Field "q" must be filled`, null);
                }
                const response = yield this.request(`search/${service}`, data, `POST`);
                return response;
            }
            catch (error) {
                return (0, logger_1.HandleError)(error, this.logger, this.formatResponse, this.search.name);
            }
        });
    }
}
exports.Search = Search;
