// @blnkfinance/blnk-typescript/dist/tests/e2e/blnk/endpoints/ledgers.e2e.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable n/no-unpublished-import */
const tap_1 = __importDefault(require("tap"));
const utils_test_1 = require("../../../utils.test");
const src_1 = require("../../../../src");
const path_1 = __importDefault(require("path"));
let ledgerId = ``;
let ledgerBalanceId = ``;
let transactionId = ``;
let identityId = ``;
let balanceMonitorId = ``;
let uploadId = ``;
let matchingRuleId = ``;
const ledgerBalanceFields = [
    `balance`,
    `version`,
    `inflight_balance`,
    `credit_balance`,
    `debit_balance`,
    `inflight_debit_balance`,
    `inflight_credit_balance`,
    `ledger_id`,
    `identity_id`,
    `balance_id`,
    `currency`,
    `created_at`,
    `inflight_expires_at`,
];
const transactionFields = [
    `transaction_id`,
    `amount`,
    `currency`,
    `source`,
    `destination`,
    `reference`,
    `status`,
    `created_at`,
    `precision`,
];
tap_1.default.test(`Ledgers end to end test`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    //   t.nock(BASE_URL).post(`/ledgers`).reply(201, {
    //     ledger_id: ledgerId,
    //     name: ``,
    //     created_at: new Date(),
    //   });
    t.test(`create a ledger`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b;
        const ledgerData = {
            name: `Test Ledger`,
        };
        const response = yield client.Ledgers.create(ledgerData);
        childTest.ok(response, `response is successful`);
        childTest.equal(response.status, 201);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [`ledger_id`, `name`, `created_at`]);
        childTest.type((_a = response.data) === null || _a === void 0 ? void 0 : _a.created_at, `string`);
        childTest.type((_b = response.data) === null || _b === void 0 ? void 0 : _b.ledger_id, `string`);
        ledgerId = response.data.ledger_id;
        childTest.end();
    }));
    t.test(`create a ledger fail`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const ledgerData = {};
        //this should fail, since we pass an empty object to the create method
        const response = yield client.Ledgers.create(ledgerData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 400);
    }));
    t.test(`It should fetch a ledger created`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Ledgers.get(ledgerId);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [`ledger_id`, `name`, `created_at`]);
        childTest.end();
    }));
    t.test(`It should return 400`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Ledgers.get(`123456789`);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.end();
}));
tap_1.default.test(`Identity`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    t.test(`create an individual identity`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const identityData = {
            category: `cutomer`,
            city: `Ikeja`,
            country: `NG`,
            first_name: `Test`,
            email_address: `test@test.com`,
            identity_type: `individual`,
            phone_number: `+2348012345678`,
            post_code: `100001`,
            state: `Lagos`,
            street: `123 Test Street`,
            dob: new Date(`1996-02-25`),
            gender: `male`,
            last_name: `Test`,
            nationality: `Nigerian`,
        };
        const response = yield client.Identity.create(identityData);
        childTest.ok(response, `identity is created`);
        childTest.equal(response.status, 201);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [
            `identity_id`,
            `created_at`,
            `identity_type`,
        ]);
        identityId = response.data.identity_id;
        childTest.end();
    }));
    t.test(`list all identities`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Identity.list();
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `array`);
        childTest.end();
    }));
    t.test(`get identity by id`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Identity.get(identityId);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [
            `identity_id`,
            `created_at`,
            `identity_type`,
        ]);
        childTest.end();
    }));
    t.test(`update identity to organization`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const identityData = {
            category: `cutomer`,
            city: `Ikeja`,
            country: `NG`,
            email_address: `test@test.com`,
            identity_type: `organization`,
            organization_name: `Test Organization`,
            phone_number: `+2348012345678`,
            post_code: `100001`,
            state: `Lagos`,
            street: `123 Test Street`,
            dob: new Date(`1996-02-25`),
            gender: `male`,
        };
        const response = yield client.Identity.update(identityId, identityData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
    }));
    t.end();
}));
tap_1.default.test(`Ledger balances`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    t.test(`It should create a balance`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const balanceData = {
            currency: `USD`,
            ledger_id: ledgerId,
            identity_id: identityId,
            meta_data: {},
        };
        const response = yield client.LedgerBalances.create(balanceData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 201);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, ledgerBalanceFields);
        ledgerBalanceId = response.data.balance_id;
    }));
    t.test(`get ledger balances`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.LedgerBalances.get(ledgerBalanceId);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, ledgerBalanceFields);
        childTest.end();
    }));
    t.test(`it should return 400`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.LedgerBalances.get(`123456789`);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.end();
}));
tap_1.default.test(`Ledger balance transactions`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    t.test(`It should create a transaction on a balance`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const transactionData = {
            amount: 1000,
            currency: `USD`,
            description: `Test transaction`,
            precision: 100,
            reference: (0, utils_test_1.GenerateRandomNumbersWithPrefix)(`test`, 4),
            source: `@bank-account`,
            destination: ledgerBalanceId,
            inflight: true,
            allow_overdraft: true, // Enable overdraft for the first deposit
            meta_data: {},
        };
        const response = yield client.Transactions.create(transactionData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 201);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [
            ...transactionFields,
            `precise_amount`,
        ]);
        transactionId = response.data.transaction_id;
        childTest.end();
    }));
    t.test(`it should commit the transaction`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        //we sleep a bit so the transaction enqueues
        yield (0, utils_test_1.Sleep)(2);
        const response = yield client.Transactions.updateStatus(transactionId, {
            status: `commit`,
        });
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [
            ...transactionFields,
            `precise_amount`,
        ]);
        childTest.end();
    }));
    t.test(`it should void the transaction`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        //we sleep a bit so the transaction enqueues
        yield (0, utils_test_1.Sleep)(2);
        const response = yield client.Transactions.updateStatus(transactionId, {
            status: `void`,
        });
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, transactionFields);
        childTest.end();
    }));
    t.end();
}));
tap_1.default.test(`Balance Monitors`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    t.test(`It should create a balance monitor`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const monitorData = {
            balance_id: ledgerBalanceId,
            condition: {
                field: `credit_balance`,
                operator: `>=`,
                precision: 100,
                value: 1000,
            },
            description: `Tier 1 account`,
        };
        const response = yield client.BalanceMonitor.create(monitorData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 201);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [`monitor_id`, `balance_id`]);
        balanceMonitorId = response.data.monitor_id;
        childTest.end();
    }));
    t.test(`It should list all balance monitors`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.BalanceMonitor.list();
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.type(response.data, `array`);
        childTest.end();
    }));
    t.test(`It should fail to get the balance monitor`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.BalanceMonitor.get(`123456789`);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 400);
        childTest.end();
    }));
    t.test(`It should get a balance monitor`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.BalanceMonitor.get(balanceMonitorId);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        // childTest.hasProps(response.data!, []);
        childTest.end();
    }));
    t.test(`It should update a balance monitor`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const monitorData = {
            balance_id: ledgerBalanceId,
            condition: {
                field: `credit_balance`,
                operator: `>=`,
                precision: 100,
                value: 1000,
            },
            description: `Tier 1 account`,
        };
        const response = yield client.BalanceMonitor.update(balanceMonitorId, monitorData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
    }));
    t.end();
}));
tap_1.default.test(`Reconciliation`, (t) => __awaiter(void 0, void 0, void 0, function* () {
    const filePath = path_1.default.join(__dirname, `..`, `..`, `..`, `..`, `file.csv`);
    const clientOptions = {
        baseUrl: utils_test_1.BASE_URL,
    };
    const client = (0, src_1.BlnkInit)(``, clientOptions);
    t.test(`It should create a matching rule`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const matchingRuleData = {
            description: `Test matching rule`,
            criteria: [
                {
                    field: `amount`,
                    operator: `equals`,
                    allowable_drift: 0.01,
                },
            ],
            name: `Test matching rule`,
        };
        const response = yield client.Reconciliation.createMatchingRule(matchingRuleData);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 201);
        matchingRuleId = response.data.rule_id;
        childTest.end();
    }));
    t.test(`it should fail to upload a reconciliation file`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Reconciliation.upload(`no path`, `Stripe`);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 404);
        childTest.end();
    }));
    t.test(`it should upload a reconciliation file`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const response = yield client.Reconciliation.upload(filePath, `Stripe`);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [`upload_id`]);
        uploadId = response.data.upload_id;
        childTest.end();
    }));
    t.test(`Run Reconciliation`, (childTest) => __awaiter(void 0, void 0, void 0, function* () {
        const data = {
            dry_run: true,
            grouping_criteria: `amount`,
            matching_rule_ids: [matchingRuleId],
            strategy: `one_to_many`,
            upload_id: uploadId,
        };
        const response = yield client.Reconciliation.run(data);
        childTest.ok(response, `response is returned`);
        childTest.equal(response.status, 200);
        childTest.type(response.data, `object`);
        childTest.hasProps(response.data, [`reconciliation_id`]);
        childTest.end();
    }));
    t.end();
}));
